[sw, tests] Introduce a test to handle two different consecutive IRQs

silvestrst opened the pull request:
The purpose of this test is to validate external interrupt delivery from peripheral to the target. The aim is not to test the entire PLIC or entirety of the peripherals, but to ensure that two different IRQs can interrupt the processor and be successfully handled. For the purpose of this test the UART peripheral is used as the origin of the interrupts, and the interrupted target is Ibex (the only target in current Earl Grey implementation).
    
This test ensures (in terms of UART RX Overflow and TX Empty IRQ sources, and Ibex target):
* That PLIC getaway receives and handles an IRQ correctly.
* That PLIC Core and Target modules handle an IRQ correctly.
* That an IRQ can interrupt the target.
* That an IRQ can be cleared and a new IRQ can be processed.
    
The reason for introducing this test was an issues we had previously with the IRQ delivery on the FPGA, see issue #1355 for more details. This test can be run by CI, and used manually to test the IRQs on an FPGA.

This PR was tested and passed on both FPGA and Verilator.
    
Fixes: #1402

silvestrst commented:
In the latest push:
* Fixed clang format

mcy commented on sw/device/tests/consecutive_irqs/consecutive_irqs_test.c:
```diff
@@ -0,0 +1,292 @@
+/**
+ * Polled implementation of the string send operation.
+ */
+static void uart_send_string(const char *string) {
```
Could this be used with print.h, or is the fact that this is polling incompatible?

mcy replied:
`print.h` doesn't know anything about UART, you have to set up a specific callback for it; how it calls UART is up to you.

mcy commented on sw/device/tests/consecutive_irqs/consecutive_irqs_test.c:
```diff
@@ -0,0 +1,292 @@
+      break;
+    default:
+      debug_msg_and_abort("Unrecognised peripheral!\n");
+  };
```
I would maybe consider something like
```c
if (claim_data.peripheral != kDifPlicPeripheralUart) {
  abort();
}
handler_uart_is(&claim_data);
```
Thoughts?

mcy commented on sw/device/tests/consecutive_irqs/consecutive_irqs_test.c:
```diff
@@ -0,0 +1,292 @@
+  return true;
+}
+
+int main(void) {
```
Should be `(int, char*)`. It's silly, but that's what the CRT calls main as, and we may as well be consistent.

mcy commented on sw/device/tests/consecutive_irqs/consecutive_irqs_test.c:
```diff
@@ -0,0 +1,292 @@
+
+  mmio_region_t uart_base_addr = mmio_region_from_addr(UART0_BASE_ADDR);
+  if (!uart_initialise(uart_base_addr, &uart0)) {
+    uart_send_string("FAIL!\n");
```
Why not `debug_message_and_abort`?

mcy commented on sw/device/tests/consecutive_irqs/consecutive_irqs_test.c:
```diff
@@ -0,0 +1,292 @@
+    uart_send_string("TEST failed to force RX overflow IRQ\n");
+    return false;
+  }
+  busy_sleep_micros(500000);
```
Is spinning actually correct here? It seems to me like you just want to `wait_for_interrupt()` but correct me if I'm wrong.

gkelly replied:
I think waiting is probably fine.

silvestrst replied:
Done.

silvestrst commented:
In the latest push:
* adressed @mcy comments.
* decreased sleep time to 10uS.

alphan commented:
> 50kHZ not 500kHZ

Hi @silvestrst, simulation clock speed is actually defined as 500 kHz in `ibex.c`, that seems to be a typo in my comment. Apologies for the confusion. But, if you want to make sure that you spin for a certain number of cycles as opposed to microseconds, writing a helper function that uses `kIbexClockFreqHz` may be a better idea:
```
/**
 * Spin for at least the given number of cycles.
 * Should be safe for |cycles| <= 4200.
 */
void cycle_sleep(uint32_t cycles) {
  uint32_t usec = cycles * 1000000 / kIbexClockFreqHz;
  usleep(usec);
}
```

Please note that `usec` will be zero if `cycles` is less than `kIbexClockFreqHz/1000000` (cycles per microsecond). You can add a `min` if needed.

gkelly commented on ci/run_verilator_pytest.sh:
```diff
@@ -20,6 +20,7 @@ TEST_TARGETS=(
   "tests/flash_ctrl/flash_test.elf"
   "tests/hmac/sha256_test.elf"
   "tests/rv_timer/rv_timer_test.elf"
+  "tests/consecutive_irqs/consecutive_irqs_test.elf"
```
Alpha sort this please.

silvestrst replied:
Done.

gkelly commented on sw/device/tests/consecutive_irqs/consecutive_irqs_test.c:
```diff
@@ -0,0 +1,289 @@
+  for (int i = 0; string[i] != '\0'; ++i) {
+    uint8_t byte = (uint8_t)string[i];
+    if (!dif_uart_byte_send_polled(&uart0, byte)) {
+      return;
```
Should this `abort()`?

silvestrst replied:
Good point, done.

Is implemented in terms of `uart_print_char` now.

gkelly commented on sw/device/tests/consecutive_irqs/consecutive_irqs_test.c:
```diff
@@ -0,0 +1,289 @@
+
+static void debug_msg_and_abort(const char *msg) {
+  uart_send_string(msg);
+  uart_send_string("FAIL!");
```
I think the harness is looking for `\n` in the PASS/FAIL message.

mcy replied:
\r\n, specifically.

silvestrst replied:
Done.

mcy commented on sw/device/tests/consecutive_irqs/consecutive_irqs_test.c:
```diff
@@ -0,0 +1,289 @@
+  return true;
+}
+
+int main(int argc, char *argv[]) {
```
`char **` is equivalent and less confusing-looking.

silvestrst replied:
Done.

mcy commented on sw/device/tests/consecutive_irqs/consecutive_irqs_test.c:
```diff
@@ -0,0 +1,289 @@
+
+int main(int argc, char *argv[]) {
+  (void)argc;
+  (void)argv;
```
You don't need to void unused function arguments.

silvestrst replied:
Done.

silvestrst commented:
I lost now the chain off comments regarding using `wait_for_interrupt` instead of a busy loop. The problem is that core will likely be interrupted before the call to `wait_for_interrupt`, and the effect of it is - core goes into power-saving-mode after the IRQ has been serviced. The result is - core doesn't wake up.

To summarise, I am keeping the busy with a slight tweak.

silvestrst commented:
In the latest push:
* I have addressed all outstanding review comments.
* Started using `print_log` for debug messages.
* Changed test logic to check whether the IRQ has been serviced, and if not then wait 10uS and 
  check again. I expect the IRQs to trigger before the check, however I am not sure whether this will 
  always be the case, and safety net of 10uS seems reasonable.
* Other minor cosmetic changes.

silvestrst commented:
Just tested this change with #1520 and some additional changes to eliminate definition clashes, and this change works as expected and passes the test. The next step will be to rebase it when #1520 lands.

alphan commented on sw/device/tests/consecutive_irqs/consecutive_irqs_test.c:
```diff
@@ -0,0 +1,292 @@
+              "PLIC: RX overflow trigger type set failed!\n");
+    return false;
+  }
+  if (!dif_plic_irq_trigger_type_set(plic, kDifUartInterruptTxEmpty,
```
`kDifUartInterruptTxEmpty` is a `dif_uart_interrupt_t` but `dif_plic_irq_trigger_type_set` expects a `dif_plic_irq_id_t`. Is this intentional?

silvestrst replied:
@alphan thank you, that is a genuine bug, I will fix it.

silvestrst replied:
I have fixed it now, but I think I will push an update after #1520 is resolved.

silvestrst commented:
In the latest push:
* Renamed `consecutive_irqs_test.elf` to `consecutive_irqs_test_sim_verilator.elf` in `run_verilator_pytest.sh`.

silvestrst commented:
In the latest push:
* I have resolved duplicated `\r`.

imphil commented on sw/device/tests/consecutive_irqs/meson.build:
```diff
@@ -0,0 +1,39 @@
+    build_always_stale: true,
+    build_by_default: true,
+  )
+endforeach
```
Please add a trailing newline.

silvestrst replied:
Oh no, they catch me out every time :) Thanks

silvestrst commented:
I think it is good to be merged, unless there are any more comments.
